### PR TITLE
fix : cast interface can handle vyper contracts

### PR DESCRIPTION
**Pull request by aliberkhsl**
## Motivation

cast interface wasn't handling vyper contracts because of missing stuff related struct types of vyper contracts such as they do not have internal_types in the abi file specially when output variable is struct

## Solution

I added missing behavior for such case that I explained in motivation but also noticed that some vyper contracts are using solidity syntax words in some variables so in that case it fails and in my solution it doesn't set output value as struct but unpack the variables directly because cause of structure of vyper contract abis it is impossible to fetch which struct it is
therefore it can't add structs to the interface as well user needs to handle it manually

Also added 2 more tests cases and confirmed all the tests are passing

### CODE REVIEW

File: utils/src/abi.rs [hunk 1]
```diff
@@ -219,14 +227,21 @@ fn format_function_output_param(
...
+            | ParamType::String
+            | ParamType::FixedArray(_, _)
+            | ParamType::Tuple(_),
     );
+    is_memory =
+        if (matches!(param.kind, ParamType::Tuple(_)) && param.internal_type.as_ref() == None) {
+            false
+        } else {
+            is_memory
+        };
```
Comment (mattsse) [inline utils/src/abi.rs:243]: this can be merged into a single match statement:

```rust
match param.kind {
   ParamType::Array(_)
            | ParamType::Bytes
            | ParamType::String
            | ParamType::FixedArray(_, _) => true,
ParamType::Tuple(_) => {...}
_ => false
}
```

File: utils/src/abi.rs [hunk 2]
```diff
@@ -182,14 +182,22 @@ fn expand_output_param_type(
             Ok(format!("{}[{}]", ty, *size))
         }
         ParamType::Tuple(_) => {
-            let ty = if let Some(struct_name) = structs
-                .get_function_output_struct_type(&fun.name, param.internal_type.as_ref().unwrap())
-            {
-                struct_name.to_string()
+            if param.internal_type.as_ref() == None {
```
Comment (mattsse) [inline utils/src/abi.rs:185]:
```suggestion
            if param.internal_type.is_none() {
```

File: utils/src/abi.rs [hunk 3]
```diff
@@ -182,14 +182,22 @@ fn expand_output_param_type(
...
+                let mut result = kind.to_string();
+                result.pop();
+                result.remove(0);
```
Comment (mattsse) [inline utils/src/abi.rs:188]: this removes first and list char.
why is this necessary here? what is the first/last char in this context?

**Review by mattsse — CHANGES_REQUESTED** [review]
thanks for this,

smol nits and one question

File: utils/src/abi.rs (hunk 3 above)
Comment (aliberkhsl) [reply utils/src/abi.rs:188]: It was adding parentheses for the struct variables so removed them but let me check again if it's really necessary, I tested generated interface in our production btw and it works

Comment (mattsse) [reply utils/src/abi.rs:188]: I think we can just `trim_start_matches`/`trim_end_matches` then?

File: utils/src/abi.rs (hunk 1 above)
Comment (aliberkhsl) [reply utils/src/abi.rs]: changed it

File: utils/src/abi.rs (hunk 3 above)
Comment (aliberkhsl) [reply utils/src/abi.rs]: Yeah changed to use it thanks for recommendation

**Comment by mattsse** [comment]
@aliberkhsl needs nightly formatting

`cargo +nightly fmt`

and there are some clippy lints that you can outfix for the `utils` crate with

```
cargo +nightly clippy --fix --all-features -Z unstable-options
```

**Comment by aliberkhsl** [comment]
Fixed the issue regarding linting now should be fine
